src/main/scala/org/constellation/primitives/TransactionGenerator.scala

```diff
@@ -0,0 +1,253 @@
+    LiftIO[F].liftIO(dao.peerInfo(NodeType.Light))
+
+  private def logTransaction(tcd: TransactionCacheData): F[Unit] =
+    Sync[F].delay { dao.miscLogger.debug(s"Rebroadcast tx=${tcd.transaction.hash}, initial path=${tcd.path}") }
+
+  private def broadcastTransaction(tcd: TransactionCacheData, peerData: List[PeerData]) = {
+    val contextShift: ContextShift[IO] = ConstellationContextShift.edge
+    LiftIO[F].liftIO(contextShift.shift *> peerData.traverse(_.client.putAsync("transaction", TransactionGossip(tcd))))
+  }
+
+  private def peerData(peers: Set[Schema.Id]): F[List[PeerData]] =
+    LiftIO[F].liftIO(dao.peerInfo(NodeType.Full).map(_.filterKeys(peers.contains).values.toList))
+
+  private def observeTransaction(transaction: Transaction): F[TransactionCacheData] =
+    transactionGossiping.observe(TransactionCacheData(transaction))
+
+  private def selectPeers(transactionCacheData: TransactionCacheData): F[Set[Schema.Id]] = {
+    implicit val random: Random = scala.util.Random
+    transactionGossiping.selectPeers(transactionCacheData)
+  }
+
+  private def putTransaction(tx: Transaction): F[TransactionCacheData] =
+    transactionService.put(TransactionCacheData(tx, path = Set(dao.id)))
+
+  private def numberOfTransaction(peers: Seq[(Id, PeerData)]): F[Int] =
+    Sync[F].pure { (dao.processingConfig.randomTXPerRoundPerPeer / peers.size + 1) + 1 }
+
+  private def keyPairForSource(sourceAddress: String): KeyPair =
+    (dao.addressToKeyPair + (dao.selfAddressStr -> dao.keyPair))(sourceAddress)
+
+  private def checkAddressesHaveSufficient(balancesAddresses: Seq[(String, Option[AddressCacheData])]): Boolean =
+    balancesAddresses.forall { _._2.exists(_.balance > requiredBalance) }
+
+  private def balancesForAddresses: F[List[(String, Option[AddressCacheData])]] =
+    dao.addresses.toList.traverse(address => addressService.lookup(address).map(address -> _))
+
+  private def getSourceAddressForTxWithoutSufficient(address: Seq[(String, Option[AddressCacheData])]) = {
+    val addresses = dao.addresses
+    val historyCheckPassable = address.forall(_._2.exists(_.balanceByLatestSnapshot > requiredBalance))
+    if (historyCheckPassable && addresses.nonEmpty) Random.shuffle(addresses :+ dao.selfAddressStr).head
+    else dao.selfAddressStr
+  }
+
+  private def randomAmount(n: Int): Long = Random.nextInt(n).toLong + 1L
+
+  private def randomAddressFromPeers(peers: Seq[(Id, PeerData)]): String =
+    if (peers.isEmpty && dao.nodeConfig.isGenesisNode) dao.dummyAddress
+    else peers(Random.nextInt(peers.size))._1.address
+
+  private def randomAddressFrom(addresses: Seq[(String, Option[AddressCacheData])]): String =
+    if (addresses.isEmpty) dao.dummyAddress
+    else Random.shuffle(addresses.filterNot(_._2.exists(_.balance > requiredBalance))).head._1
+
+  private def getReadyPeers: F[Seq[(Id, PeerData)]] = {
+    val timeMillis = System.currentTimeMillis() - (dao.processingConfig.minPeerTimeAddedSeconds * 1000)
```

Comment:
Maybe it would be better to embed it within the method call, as there is potential risk that this will be called ahead of time when outside of an `F`